### PR TITLE
Publish public preview swagger spec for "Azure metrics Ingestion REST  API"

### PR DESCRIPTION
**Pull request by vickyv51**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [Done] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
Missing README.md file, see [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/monitor/resource-manager/readme.md) for an example.

File: specification/monitor/data-plane/preview/2018-09-01-preview.json [hunk 1]
```diff
@@ -0,0 +1,1927 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "The azure metrics publish succeeded",
+            "schema": {
+              "$ref": "#/definitions/Microsoft.AspNetCore.Mvc.ObjectResult"
+            }
+          },
+          "400": {
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/preview/2018-09-01-preview.json]: Shouldn't need to model these failure cases, use ```default``` instead.  See [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/datalake-store/data-plane/Microsoft.DataLakeStore/stable/2016-11-01/filesystem.json#L95) for an example.
Also is it expected that the 200 and 4xx cases return the same model?

File: specification/monitor/data-plane/preview/2018-09-01-preview.json [hunk 2]
```diff
@@ -0,0 +1,1927 @@
...
+    "application/json; odata=minimalmetadata"
+  ],
+  "produces": [
+    "application/json"
+  ],
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{resourceProvider}/{resourceTypeName}/{resourceName}/metrics": {
+      "post": {
+        "summary": "Ingestion API used by Azure custom metrics\r\n            Responsible of below operation\r\n            1. Authorize the request\r\n            2. Deserialize payload request body\r\n            3. Perform validation of the payload\r\n            4. Commits the payload for metrics ingestion",
+        "operationId": "postSubscriptionsBySubscriptionIdResourcegroupsByResourceGroupNameProvidersByResourceProviderByResourceTypeNameByResourceNameMetrics",
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/preview/2018-09-01-preview.json]: The ```operationId``` is used to generate method names. Please shorten it to something more descriptive.

Comment (jhendrixMSFT) [inline specification/monitor/data-plane/preview/2018-09-01-preview.json]: Please provide an example for this operation, see [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/datalake-store/data-plane/Microsoft.DataLakeStore/stable/2016-11-01/filesystem.json#L102) for an example.

File: specification/monitor/data-plane/preview/2018-09-01-preview.json (hunk 1 above)
Comment (jhendrixMSFT) [reply specification/monitor/data-plane/preview/2018-09-01-preview.json]: Ideally this information would be returned in the response body as an [OData v4 object](http://docs.oasis-open.org/odata/odata-json-format/v4.0/os/odata-json-format-v4.0-os.html#_Toc372793091).  Something like this.
```json
{
  "Error": {
    "Code": "InvalidRequest",
    "Message": "Some descriptive error message."
  }
}
```

Comment (vickyv51) [reply specification/monitor/data-plane/preview/2018-09-01-preview.json]: We indeed already do that. So is the suggestion that having above behavior good enough?

Comment (jhendrixMSFT) [reply specification/monitor/data-plane/preview/2018-09-01-preview.json]: I would suggest then that you specify a ```default``` response that references the appropriate model definition.  Looking at the definition for ```Microsoft.AspNetCore.Mvc.ObjectResult``` it's not an OData v4 error object (or a superset of it).

Comment (jhendrixMSFT) [reply specification/monitor/data-plane/preview/2018-09-01-preview.json]: More importantly, the code generators will treat any status code modeled here as a success.  So if these 4xx codes aren't to be treated as success then you will want to remove them.

Comment (vickyv51) [reply specification/monitor/data-plane/preview/2018-09-01-preview.json]: Makes sense. Fixing it. :)

File: specification/monitor/data-plane/preview/2018-09-01-preview.json (hunk 2 above)
Comment (vickyv51) [reply specification/monitor/data-plane/preview/2018-09-01-preview.json]: Updating in next iteration

File: specification/monitor/data-plane/readme.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+``` yaml $(java)
+azure-arm: true
+fluent: true
+namespace: com.microsoft.azure.management.monitor
+license-header: MICROSOFT_MIT_NO_CODEGEN
+payload-flattening-threshold: 1
+output-folder: $(azure-libraries-for-java-folder)/azure-mgmt-monitor
+```
+
+# Validation
```
[the remark this answers is not in the document]
Comment (vickyv51) [reply specification/monitor/data-plane/readme.md]: Thats correct. Dropping it.

File: specification/monitor/data-plane/readme.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+tag: package-2018-09
+
+```
+### Tag: package-2018-09
+
+These settings apply only when `--tag=package-2018-09` is specified on the command line.
+
+``` yaml $(tag) == 'package-2018-09'
+input-file:
+- data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: This path is relative to the location of the README.md file, it should be `preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json`.

File: specification/monitor/data-plane/readme.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+title: MonitorClient
+description: Monitor Management Client
+openapi-type: arm
+
+tag: package-2018-09
+
+```
+### Tag: package-2018-09
+
+These settings apply only when `--tag=package-2018-09` is specified on the command line.
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: I would add a `-preview` suffix to the tag name so it's clear this is preview, plus if you end up releasing a 2018-09-01 GA the tag names will be unique.

File: specification/monitor/data-plane/readme.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+These settings apply only when `--csharp` is specified on the command line.
+Please also specify `--csharp-sdks-folder=<path to "SDKs" directory of your azure-sdk-for-net clone>`.
+
+``` yaml $(csharp)
+csharp:
+  azure-arm: true
+  payload-flattening-threshold: 1
+  license-header: MICROSOFT_MIT_NO_VERSION
+  namespace: Microsoft.Azure.Management.Monitor
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: Is this the correct namespace for the C# implementation?  @shahabhijeet can you review this part?

File: specification/monitor/data-plane/readme.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+### Tag: package-2018-09 and go
+
+These settings apply only when `--tag=package-2018-09 --go` is specified on the command line.
+Please also specify `--go-sdk-folder=<path to the root directory of your azure-sdk-for-go clone>`.
+
+``` yaml $(tag) == 'package-2018-09' && $(go)
+output-folder: $(go-sdk-folder)/services/preview/monitor/mgmt/2018-03-01/insights
+```
+
+## Python
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: @lmazuel can you please review the config section for python?

File: specification/monitor/data-plane/readme.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  no-namespace-folders: true
+  output-folder: $(python-sdks-folder)/azure-mgmt-monitor/azure/mgmt/monitor
+```
+``` yaml $(python) && $(python-mode) == 'create'
+python:
+  basic-setup-py: true
+  output-folder: $(python-sdks-folder)/azure-mgmt-monitor
+```
+
+## Java
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: @jianghaolu can you please review the config section for Java?

File: specification/monitor/data-plane/readme.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+
+
+### Basic Information
+These are the global settings for the MonitorClient API.
+
+``` yaml
+title: MonitorClient
+description: Monitor Management Client
+openapi-type: arm
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: The value here should be `data-plane` not `arm`.

File: specification/monitor/data-plane/readme.md (hunk 6 above)
Comment (jhendrixMSFT) [reply specification/monitor/data-plane/readme.md]: Given that this is data-plane I would think the word `Management` doesn't belong in the namespace but I'm not really sure.

File: specification/monitor/data-plane/readme.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+  - tag: package-2018-09
+```
+
+### Tag: package-2018-09 and go
+
+These settings apply only when `--tag=package-2018-09 --go` is specified on the command line.
+Please also specify `--go-sdk-folder=<path to the root directory of your azure-sdk-for-go clone>`.
+
+``` yaml $(tag) == 'package-2018-09' && $(go)
+output-folder: $(go-sdk-folder)/services/preview/monitor/mgmt/2018-03-01/insights
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: Value is wrong, this should be `$(go-sdk-folder)/services/preview/monitor/2018-09-01-preview/monitor`

File: specification/monitor/data-plane/readme.md (hunk 4 above)
Comment (vickyv51) [reply specification/monitor/data-plane/readme.md]: My bad. Thanks for catching it. Fixed

File: specification/monitor/data-plane/readme.md (hunk 5 above)
Comment (vickyv51) [reply specification/monitor/data-plane/readme.md]: Makes sense. Fixing it.

File: specification/monitor/data-plane/readme.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+
+## Go
+
+These settings apply only when `--go` is specified on the command line.
+
+``` yaml $(go)
+go:
+  license-header: MICROSOFT_APACHE_NO_VERSION
+  namespace: insights
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: If the package directory is going to be `monitor` then this namespace will need to match it.

File: specification/monitor/data-plane/readme.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+
+## Go
+
+These settings apply only when `--go` is specified on the command line.
+
+``` yaml $(go)
+go:
+  license-header: MICROSOFT_APACHE_NO_VERSION
+  namespace: azuremonitor
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/readme.md]: This needs to match the package directory (i.e. the leaf directory on line 104).

**Review by shahabhijeet — APPROVED** [review]

File: specification/monitor/data-plane/readme.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+Please also specify `--python-sdks-folder=<path to the root directory of your azure-sdk-for-python clone>`.
+Use `--python-mode=update` if you already have a setup.py and just want to update the code itself.
+
+``` yaml $(python)
+python-mode: create
+python:
+  azure-arm: true
+  license-header: MICROSOFT_MIT_NO_VERSION
+  payload-flattening-threshold: 2
+  namespace: microsoft.azure.monitor.metrics
```
Comment (lmazuel) [inline specification/monitor/data-plane/readme.md]: azure.monitor

File: specification/monitor/data-plane/readme.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+  package-name: microsoft.azure.monitor.metrics
```
Comment (lmazuel) [inline specification/monitor/data-plane/readme.md]: azure-monitor

File: specification/monitor/data-plane/readme.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+  package-version: 0.5.0
+  clear-output-folder: true
+```
+``` yaml $(python) && $(python-mode) == 'update'
+python:
+  no-namespace-folders: true
+  output-folder: $(python-sdks-folder)/azure-monitor/azure/dataplane/monitor
```
Comment (lmazuel) [inline specification/monitor/data-plane/readme.md]: output-folder: $(python-sdks-folder)/azure-monitor/azure/monitor

File: specification/monitor/data-plane/readme.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+```
+``` yaml $(python) && $(python-mode) == 'create'
+python:
+  basic-setup-py: true
+  output-folder: $(python-sdks-folder)/azure-monitor/azure/dataplane/monitor
```
Comment (lmazuel) [inline specification/monitor/data-plane/readme.md]: output-folder: $(python-sdks-folder)/azure-monitor/

File: specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json [hunk 17]
```diff
@@ -0,0 +1,1896 @@
...
+            }
+          },
+          "default": {
+            "description": "An unexpected error from the server. See response object for the reason",
+            "schema": {
+              "$ref": "#/definitions/Microsoft.AspNetCore.Mvc.ObjectResult"
+            }
+          }
+        },
+        "x-ms-example": {
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json]: Should be `x-ms-examples`

File: specification/monitor/data-plane/readme.md (hunk 14 above)
Comment (vickyv51) [reply specification/monitor/data-plane/readme.md]: I'm fixing this assuming below
namespace: azure.monitor
package-name: azure-monitor

File: specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json [hunk 18]
```diff
@@ -0,0 +1,1896 @@
...
+          "Post azure metrics for cows sold with related metadata as dimensions": {
+            "$ref": "./examples/AzureMetrics_IngestMetrics.json"
+          }
+        }
+      }
+    }
+  },
+  "definitions": {
+    "Microsoft.Azure.Monitoring.Ingestion.Gateway.Models.AzureMetricsDocument": {
```
[the remark this answers is not in the document]
Comment (vickyv51) [reply specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json]: These are actually generated directly using swagger extension based on the code using VS annotations. I will drop the prefix to make it more non c#-ish

File: specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json [hunk 19]
```diff
@@ -0,0 +1,1896 @@
...
+          "type": "number"
+        },
+        "Count": {
+          "format": "int32",
+          "description": "Gets or sets Count value",
+          "type": "integer"
+        }
+      }
+    },
+    "Microsoft.AspNetCore.Mvc.ObjectResult": {
```
Comment (jhendrixMSFT) [inline specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json]: More prefix to remove, also on line 208.

**Comment by jhendrixMSFT** [comment]
Please take a look at the [linter](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/427443386) errors and [model validation](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/427443392) failures.

File: specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json [hunk 20]
```diff
@@ -0,0 +1,1896 @@
...
+        "DeclaredType": {
+          "$ref": "#/definitions/System.Type"
+        },
+        "StatusCode": {
+          "format": "int32",
+          "type": "integer"
+        }
+      }
+    },
+    "Microsoft.AspNetCore.Mvc.Formatters.IOutputFormatter": {
```
[the remark this answers is not in the document]
Comment (vickyv51) [reply specification/monitor/data-plane/preview/2018-09-01-preview/azureMonitorCustomMetricsIngestionApi.json]: Fixed.

File: specification/monitor/data-plane/readme.md [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+
+## Java
+
+These settings apply only when `--java` is specified on the command line.
+Please also specify `--azure-libraries-for-java-folder=<path to the root directory of your azure-libraries-for-java clone>`.
+
+``` yaml $(java)
+azure-arm: true
+fluent: true
+namespace: com.microsoft.azure.dataplane.monitor
```
Comment (jianghaolu) [inline specification/monitor/data-plane/readme.md]: simply "com.microsoft.azure.monitor"

File: specification/monitor/data-plane/readme.md [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+license-header: MICROSOFT_MIT_NO_CODEGEN
+payload-flattening-threshold: 1
+output-folder: $(azure-libraries-for-java-folder)/azure-monitor
```
Comment (jianghaolu) [inline specification/monitor/data-plane/readme.md]: $(azure-libraries-for-java-folder)/monitor/data-plane

**Review by jianghaolu — CHANGES_REQUESTED** [review]

**Comment by jhendrixMSFT** [comment]
There are multiple enum types named `attributes` that have different sets of values which is not allowed, see [this](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/427946630) build log.  Either they all need to have the same set of values (which in that case just define it once and reference it) or their names need to be different.
Also you can run some of these validation tools locally, see https://github.com/Azure/oav for details.

**Comment by jhendrixMSFT** [comment]
Down to just two errors in the model validation regarding the example, see the [log](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/428294649).

**Review by jhendrixMSFT — APPROVED** [review]

**Review by jianghaolu — APPROVED** [review]

**Pull request merged**